pandas/_libs/lib.pyx

```diff
@@ -793,14 +793,16 @@ def count_level_2d(ndarray[uint8_t, ndim=2, cast=True] mask,
         with nogil:
             for i in range(n):
                 for j in range(k):
-                    counts[labels[i], j] += mask[i, j]
+                    if mask[i, j]:
+                        counts[labels[i], j] += 1
```

Comment:
Sorry if missed before but can you share why this change is required? Isn't the mask only ever 0 or 1?

Reply:
The segfault is caused by out-of-bounds access on `counts`, when `labels[i]` is `-1`. If it is checked first whether a value should be counted then out-of-bounds access does not happen.

This is also how the code was written originally:
https://github.com/pandas-dev/pandas/commit/d968aab3da13f68fabbb05f8eba12927d8bc5836#diff-8fa7422077eecd07a4006c11c78fc93aL1265-L1268

That means that `count_level_2d` requires `mask` to be `0` 
1. on columns/rows that should not be counted, because the counted-over row index or column index value is `NaN`.
1. for data values that are `NaN`

Reply:
Ah OK I see - thanks for clarifying